### PR TITLE
Fix the bug - App crashed when  Android permission request got interrupted 

### PR DESCRIPTION
**Pull request by pureklkl**
The code to handle the request result in PermissionAndorid didn't have bound check on grand results.

When a request process was interrupted, the result will have zero length and cause the app crash.

I fixed it by returning a null for interrupted request because I don't want to change the API. But returning a Integer instead of a Boolean which was also used in Android API will be better.

The bug can be reproduced by emit two permission requests subsequently without waiting the result. The second one will interrupt the first one. App will crash on real divice (I tested it on oneplus 3, API23).
The bug seems won't happen on emulators.

The interrupted permission request was mentioned at [there](https://developer.android.com/reference/android/support/v4/app/ActivityCompat.OnRequestPermissionsResultCallback.html)

> "Note: It is possible that the permissions request interaction with the user is interrupted. In this case you will receive empty permissions and results arrays which should be treated as a cancellation."

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 1]
```diff
@@ -116,7 +119,10 @@ public boolean onRequestPermissionsResult(
       int requestCode,
       String[] permissions,
       int[] grantResults) {
-    mCallbacks.get(requestCode).invoke(grantResults[0]);
+    if(grantResults.length>0)
```
Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: missing space around operators

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 2]
```diff
@@ -99,7 +99,10 @@ public void requestPermission(final String permission, final Promise promise) {
         mRequestCode, new Callback() {
           @Override
           public void invoke(Object... args) {
-            promise.resolve(args[0].equals(PackageManager.PERMISSION_GRANTED));
+            if(args!=null)
```
Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: missing space around operators

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 3]
```diff
@@ -99,7 +99,10 @@ public void requestPermission(final String permission, final Promise promise) {
...
+              promise.resolve(args[0].equals(PackageManager.PERMISSION_GRANTED));
+            else
+              promise.resolve(null);
```
Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: Why not call `promise.reject`?

Comment (pureklkl) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: Sure, I have changed it to be a rejection.

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 4]
```diff
@@ -31,6 +31,8 @@
 @ReactModule(name = "PermissionsAndroid")
 public class PermissionsModule extends ReactContextBaseJavaModule implements PermissionListener {
 
+  private static final String ERROR_REQUST_INTERRIPTED = "E_REQUEST_INTERRUPTED";
```
Comment (cmcewen) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: typo - "INTERRIPTED" should be "INTERRUPTED"

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 5]
```diff
@@ -99,7 +101,10 @@ public void requestPermission(final String permission, final Promise promise) {
         mRequestCode, new Callback() {
           @Override
           public void invoke(Object... args) {
-            promise.resolve(args[0].equals(PackageManager.PERMISSION_GRANTED));
+            if(args[0] != null)
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: Please add curly braces and space after `if`

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 6]
```diff
@@ -116,7 +121,10 @@ public boolean onRequestPermissionsResult(
       int requestCode,
       String[] permissions,
       int[] grantResults) {
-    mCallbacks.get(requestCode).invoke(grantResults[0]);
+    if(grantResults.length > 0)
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: same as previous comment

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java.orig [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+import com.facebook.react.modules.core.PermissionAwareActivity;
+import com.facebook.react.modules.core.PermissionListener;
+
+/**
+ * Module that exposes the Android M Permission system to JS.
+ */
+@ReactModule(name = "PermissionsAndroid")
+public class PermissionsModule extends ReactContextBaseJavaModule implements PermissionListener {
+
+<<<<<<< HEAD
```
Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java.orig]: 👎

**Comment by leeight** [comment]
PLEASE REMOVE THE `PermissionsModule.java.orig` FILE.

**Comment by cmcewen** [comment]
@satya164 this looks good to me now - what do you think?

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 8]
```diff
@@ -107,7 +109,12 @@ public void requestPermission(final String permission, final Promise promise) {
         mRequestCode, new Callback() {
           @Override
           public void invoke(Object... args) {
-            promise.resolve(args[0].equals(PackageManager.PERMISSION_GRANTED));
+            if (args[0] != null) {
+              promise.resolve(args[0].equals(PackageManager.PERMISSION_GRANTED));
+            }
+            else {
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: `}` and `else` on the same line please

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 9]
```diff
@@ -126,7 +133,12 @@ public boolean onRequestPermissionsResult(
       int requestCode,
       String[] permissions,
       int[] grantResults) {
-    mCallbacks.get(requestCode).invoke(grantResults[0]);
+    if (grantResults.length > 0) {
+      mCallbacks.get(requestCode).invoke(grantResults[0]);
+    }
+    else {
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: `}` and `else` on the same line please

File: ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java [hunk 10]
```diff
@@ -32,6 +32,8 @@
 public class PermissionsModule extends ReactContextBaseJavaModule implements PermissionListener {
 
   private static final String ERROR_INVALID_ACTIVITY = "E_INVALID_ACTIVITY";
+  private static final String ERROR_REQUST_INTERRUPTED = "E_REQUEST_INTERRUPTED";
```
Comment (lacker) [inline ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java]: REQUST is misspelled

**Comment by mkonicek** [comment]
This pull request has been inactive for a while and the last comments are questions or there have been changes requested.

I'll close the pull request so it doesn't stay open indefinitely. If you plan to continue working on this please send a new pull request.

**Pull request closed without merging**